### PR TITLE
feat: percentage increase feature

### PR DESCRIPTION
**Pull request by salimtb**
## **Description**

This pull request introduces a new feature to the MetaMask extension that enhances user experience by displaying the percentage increase or decrease for each token directly within the UI. This update aims to provide users with immediate visual feedback on the performance of their tokens, helping them make more informed decisions based on recent market trends.

core PR: https://github.com/MetaMask/core/pull/4206

[Open in GitHub Codespaces](https://codespaces.new/MetaMask/metamask-extension/pull/24223?quickstart=1)

## **Related issues**

Fixes:

## **Manual testing steps**

1. Go to the wallet view
2. You should see the percentage of increase/decrease

### **Before**
Screenshot 2024-04-24 at 15 03 39

### **After**

Screenshot 2024-04-24 at 14 53 25

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

File: ui/components/multichain/token-list-item/price/percentage-change.tsx [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+  value: number | null | undefined;
+  valueChange?: number | null | undefined;
+  includeNumber?: boolean;
+}) => {
+  const fiatCurrency = useSelector(getCurrentCurrency);
+  const locale = useSelector(getIntlLocale);
+
+  let color = TextColor.textDefault;
+
+  if (value !== null && value !== undefined && !Number.isNaN(value)) {
```
Comment (darkwing) [inline ui/components/multichain/token-list-item/price/percentage-change.tsx:73]: Can this be:

```
if(Boolean(value)) {
``

File: ui/components/multichain/token-list-item/price/percentage-change.tsx [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    if (value >= 0) {
+      color = TextColor.successDefault;
+    } else {
+      color = TextColor.errorDefault;
+    }
+  }
+
+  const formattedValue =
+    value !== null && value !== undefined && !Number.isNaN(value)
```
Comment (darkwing) [inline ui/components/multichain/token-list-item/price/percentage-change.tsx:82]: This line is repeated, btw

File: ui/components/multichain/token-list-item/price/percentage-change.tsx [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+      ? `${value >= 0 ? '+' : ''}${value.toFixed(2)}%`
```
Comment (darkwing) [inline ui/components/multichain/token-list-item/price/percentage-change.tsx:83]: Do we want a `-` for the negative case?

File: ui/components/multichain/token-list-item/price/percentage-change.tsx [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+      : '';
+
+  const formattedValuePrice =
+    valueChange !== null &&
+    valueChange !== undefined &&
+    !Number.isNaN(valueChange)
```
Comment (darkwing) [inline ui/components/multichain/token-list-item/price/percentage-change.tsx:89]: This is the third spot we use this set of `if` conditions, we should consolidate.

File: ui/components/multichain/token-list-item/token-list-item.js [hunk 5]
```diff
@@ -87,6 +91,11 @@ export const TokenListItem = ({
   const isFullScreen = environmentType === ENVIRONMENT_TYPE_FULLSCREEN;
   const history = useHistory();
 
+  const tokenPercentChange1d = useSelector(getTokenPercentChange1d);
+
+  const tokenAddress = address ? address.toLowerCase() : null;
```
Comment (darkwing) [inline ui/components/multichain/token-list-item/token-list-item.js:96]: Could this be `address?.toLowerCase()`?

**Review by darkwing — DISMISSED** [review]

File: .yarn/patches/@metamask-assets-controllers-patch-55c0923b82.patch [hunk 6]
```diff
@@ -0,0 +1,2243 @@
+diff --git a/dist/TokenRatesController.d.ts b/dist/TokenRatesController.d.ts
```
Comment (legobeat) [inline .yarn/patches/@metamask-assets-controllers-patch-55c0923b82.patch:1]: Does this patch to `@metamask/assets-controller` have corresponding PR(s) in https://github.com/metamask/core? Is there a reason to still prefer patching in all the changes here, rather than upgrading from version 26?

Comment (legobeat) [reply .yarn/patches/@metamask-assets-controllers-patch-55c0923b82.patch:1]: Related:
- #24333
- #24285
- #24261

File: ui/components/multichain/token-list-item/price/percentage-change.tsx (hunk 3 above)
Comment (salimtb) [reply ui/components/multichain/token-list-item/price/percentage-change.tsx:83]: the `-` sign is already in the state, here's an example: 
```
priceChange1d: -0.00164019539019189
pricePercentChange1d: -0.16399898596690138
```

File: .yarn/patches/@metamask-assets-controllers-patch-55c0923b82.patch (hunk 6 above)
Comment (salimtb) [reply .yarn/patches/@metamask-assets-controllers-patch-55c0923b82.patch:1]: hey @legobeat ,
thnx for your comment, you can find [here](https://github.com/MetaMask/core/pull/4206) the core PR, i also added it to the PR description.
we use a patch to ship this asap feature without having to bump the assets-controller to the latest version.
once a new release of the assets controller has been made ( core PR merged ), we'll remove the patch and bump the version.

File: ui/components/multichain/token-list-item/token-list-item.js [hunk 7]
```diff
@@ -224,10 +234,10 @@ export const TokenListItem = ({
                   >
                     {isStakeable ? (
                       <>
-                        {tokenSymbol} {stakeableTitle}
+                        {tokenTitle} {stakeableTitle}
                       </>
                     ) : (
-                      tokenSymbol
+                      tokenTitle
                     )}
```
Comment (sahar-fehri) [inline ui/components/multichain/token-list-item/token-list-item.js:241]: Quick question, why are we using tokenTitle instead of tokenSymbol?

File: test/data/mock-state.json [hunk 8]
```diff
@@ -683,11 +683,73 @@
         "decimals": "18"
       }
     ],
-    "contractExchangeRates": {
```
Comment (sahar-fehri) [inline test/data/mock-state.json:686]: 👋  The contractExchangeRates has been removed and replaced by marketData in a lot of places, does this need a migration?

File: ui/selectors/confirm-transaction.js [hunk 9]
```diff
@@ -204,12 +214,15 @@ export const sendTokenTokenAmountAndToAddressSelector = createSelector(
...
   contractExchangeRatesSelector,
   tokenAddressSelector,
-  (contractExchangeRates, tokenAddress) =>
-    contractExchangeRates[
-      Object.keys(contractExchangeRates).find((address) =>
-        isEqualCaseInsensitive(address, tokenAddress),
-      )
-    ],
+  (contractExchangeRates, tokenAddress) => {
+    console.log('HERE +++++++++++', contractExchangeRates);
```
Comment (sahar-fehri) [inline ui/selectors/confirm-transaction.js:218]: small HERE log :p

File: ui/selectors/confirm-transaction.js [hunk 10]
```diff
@@ -204,12 +214,15 @@ export const sendTokenTokenAmountAndToAddressSelector = createSelector(
...
+    return contractExchangeRates[
+      Object.keys(contractExchangeRates).find((address) => {
+        console.log('HERE 2222 -----', tokenAddress);
```
Comment (sahar-fehri) [inline ui/selectors/confirm-transaction.js:221]: And HERE too :D

File: ui/selectors/confirm-transaction.js (hunk 9 above)
Comment (salimtb) [reply ui/selectors/confirm-transaction.js]: fixed

File: test/data/mock-state.json (hunk 8 above)
Comment (salimtb) [reply test/data/mock-state.json:686]: i believe a Redux state migration won't be necessary for this case, given that the marketData property encompasses not only contractExchangeRates but also a variety of other elements. Additionally, the structure of marketData is distinct due to its aggregation by chainId. Typically, migrations are beneficial when modifications to the existing data are needed to accommodate a new structure. However, in this instance, there's a significant amount of data absent that would require fetching via an API call to complete the transition.

File: ui/selectors/selectors.js [hunk 11]
```diff
@@ -495,8 +495,24 @@ export function getTargetAccount(state, targetAddress) {
...
 }
 
-export const getTokenExchangeRates = (state) =>
-  state.metamask.contractExchangeRates;
+export const getTokenExchangeRates = (state) => {
+  const chainId = getCurrentChainId(state);
+  console.log('Here ....', state.metamask);
+  const contractMarketData = state.metamask.marketData?.[chainId] ?? {};
+
+  return Object.entries(contractMarketData).reduce(
```
Comment (bergeron) [inline ui/selectors/selectors.js:503]: This selector may need to be memoized with `createDeepEqualSelector` or similar so things don't re-render unnecessarily.

I think this is because `useSelector` uses `===` for equality, and `Object.entries` is going to create a new object each time.

File: .storybook/test-data.js [hunk 12]
```diff
@@ -467,8 +467,28 @@ const state = {
         isEns: true,
       },
     ],
-    contractExchangeRates: {
-      '0xaD6D458402F60fD3Bd25163575031ACDce07538D': 0,
+    marketData: {
```
Comment (bergeron) [inline .storybook/test-data.js:470]: Given how much of this PR relates to the breaking changes in token rate controller state (now published as assets-controller version 30), it might be cleaner to do a dedicated PR for that.

File: ui/components/multichain/token-list-item/token-list-item.js [hunk 13]
```diff
@@ -246,21 +255,25 @@ export const TokenListItem = ({
...
-                {tokenTitle}
-              </Text>
+
+              <PercentageChange
+                value={
+                  isNativeCurrency
+                    ? tokensMarketData?.[zeroAddress()]?.pricePercentChange1d
+                    : tokenPercentageChange
+                }
+                // data-testid={`percentage-change-${
```
Comment (darkwing) [inline ui/components/multichain/token-list-item/token-list-item.js:272]: Needs removal comment

File: ui/components/multichain/token-list-item/price/percentage-change/percentage-change.tsx [hunk 14]
```diff
@@ -0,0 +1,49 @@
+import React from 'react';
+import { Box, Text } from '../../../../component-library';
+import {
+  Display,
+  FontWeight,
+  TextColor,
+  TextVariant,
+} from '../../../../../helpers/constants/design-system';
+
+const isValidAmount = (amount: number | null | undefined): boolean =>
```
Comment (darkwing) [inline ui/components/multichain/token-list-item/price/percentage-change/percentage-change.tsx:10]: This same function is in multiple files ; can we add a helper file and import from that one place?

File: ui/components/multichain/token-list-item/price/percentage-and-amount-change/percentage-and-amount-change.tsx [hunk 15]
```diff
@@ -0,0 +1,122 @@
...
+  const locale = useSelector(getIntlLocale);
+  const balanceValue = useSelector(getSelectedAccountCachedBalance);
+  const conversionRate = useSelector(getConversionRate);
+  const nativeCurrency = useSelector(getNativeCurrency);
+  const marketData = useSelector(getTokensMarketData);
+
+  const balanceChange = useMemo(() => {
+    const percentage1d = marketData?.[zeroAddress()]?.pricePercentChange1d;
+
+    if (isHexString(balanceValue)) {
```
Comment (darkwing) [inline ui/components/multichain/token-list-item/price/percentage-and-amount-change/percentage-and-amount-change.tsx:70]: Can we add some comments for the math usage here so maintainers feel comfortable knowing what's happening?

**Review by darkwing — DISMISSED** [review]
Some nits but overall looks great!

File: ui/components/multichain/token-list-item/price/percentage-and-amount-change/percentage-and-amount-change.tsx (hunk 15 above)
Comment (salimtb) [reply ui/components/multichain/token-list-item/price/percentage-and-amount-change/percentage-and-amount-change.tsx:73]: done , just added some comments

File: ui/components/multichain/token-list-item/price/percentage-change/percentage-change.tsx [hunk 16]
```diff
@@ -0,0 +1,47 @@
...
+  value,
+  address,
+}: {
+  value: number | null | undefined;
+  address: `0x${string}`;
+}) => {
+  let color = TextColor.textDefault;
+
+  if (isValidAmount(value)) {
+    if ((value as number) >= 0) {
```
Comment (darkwing) [inline ui/components/multichain/token-list-item/price/percentage-change/percentage-change.tsx:21]: Non-blocking idea: in the very rare case that the percentage change is `0`, we should maybe use a neutral color.

File: ui/components/multichain/token-list-item/price/percentage-change/percentage-change.tsx [hunk 17]
```diff
@@ -0,0 +1,47 @@
...
+      color = TextColor.successDefault;
+    } else {
+      color = TextColor.errorDefault;
+    }
+  }
+
+  const formattedValue = isValidAmount(value)
+    ? `${(value as number) >= 0 ? '+' : ''}${(value as number).toFixed(2)}%`
```
Comment (darkwing) [inline ui/components/multichain/token-list-item/price/percentage-change/percentage-change.tsx:29]: Is this duplicated in `ui/components/multichain/token-list-item/price/percentage-and-amount-change/percentage-and-amount-change.tsx`?

**Review by bergeron — APPROVED** [review]
Looks good to me.  Tested some different chains, and tokens with/without prices.

I noticed the false positive scam warning on linea, but that's fixed by https://github.com/MetaMask/metamask-extension/pull/25436

**Review by sahar-fehri — APPROVED** [review]

**Comment by EtherWizard33** [comment]
nice feature
code LGTM

**Review by EtherWizard33 — APPROVED** [review]
